api/src/main/java/jakarta/enterprise/inject/spi/BeanContainer.java

```diff
@@ -251,4 +252,40 @@ public interface BeanContainer {
      * @since 2.0
      */
     Instance<Object> createInstance();
+
+    /**
+     * Facilitates programmatic access to observer method resolution rules (assignability rules).
+     * Verifies if an observer method represented by its types and qualifiers ({@link TypesAndQualifiers}) would get
+     * notified for an event with provided type {@code T} and {@code qualifiers}.
+     * <p>
+     * Users are expected to supply all qualifiers, including implied ones such as {@code @Default}.
+     * <p>
+     * Throws {@link IllegalArgumentException} if any of the arguments is {@code null}.
+     *
+     * @param observerMethod Represents the type and qualifiers of an observer method
+     * @param eventPayloadType Event payload type
+     * @param qualifiers Event qualifiers
+     * @return true if observer method with supplied {@link TypesAndQualifiers} would be notified for given payload type
+     * and qualifiers; false otherwise
+     * @param <T> represents event payload type
+     */
+    <T> boolean isMatchingObserverMethod(TypesAndQualifiers observerMethod, T eventPayloadType, Annotation... qualifiers);
+
+    /**
+     * Facilitates programmatic access to typesafe resolution rules (assignability rules).
+     * Verifies if a bean represented by its types and qualifiers ({@link TypesAndQualifiers}) matches an injection
+     * point based on required {@code injectionPointType} and {@code qualifiers}.
+     * <p>
+     * Users are expected to supply all qualifiers, including implied ones such as {@code @Default}.
+     * <p>
+     * Throws {@link IllegalArgumentException} if any of the arguments is {@code null}.
+     *
+     * @param bean Represent the type and qualifiers of a bean
+     * @param injectionPointType Type of injection point
+     * @param qualifiers Injection point qualifiers
+     * @return true if a bean with supplied {@link TypesAndQualifiers} is assignable to an injection point of given
+     * type and qualifiers; false otherwise
+     * @param <T> represents injection point type
+     */
+    <T> boolean isMatchingBean(TypesAndQualifiers bean, T injectionPointType, Annotation... qualifiers);
```

Comment:
The type parameter here is practically useless. You can pass a `String` as the parameter and noone will complain.

We could have 2 variants of the method, one taking `Class<T>` and the other taking `TypeLiteral<T>`. Or we could just use `TypeAndQualifiers` for both inputs.

Which brings me to the other idea: why is it `TypesAndQualifiers`? I believe most usages would need exactly 1 type, no?